drivers/power/ltc4162l/ltc4162l.c

```diff
@@ -0,0 +1,829 @@
+ */
+int ltc4162l_get_icharge_setting(struct ltc4162l_desc *dev, uint16_t *val)
+{
+	uint16_t regval;
+	int ret;
+
+	ret = ltc4162l_reg_read(dev, LTC4162L_CHARGE_CURRENT_SETTING, &regval);
+	if (ret)
+		return ret;
+
+	regval = no_os_field_get(LTC4162L_ICHARGE_DAC_MASK, regval);
+
+	/* The charge current servo level: (icharge_dac + 1) × 1mV/RSNSB */
+	++regval;
+
+	*val = 10000u * LTC4162L_MULT_FRAC(regval, 100000u, dev->rsnsb);
+
+	return 0;
+}
+
+/**
+ * @brief Set icharge setting
+ * @param dev - The LTC4162L device structure.
+ * @param val - The icharge setting.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_set_icharge_setting(struct ltc4162l_desc *dev, uint16_t val)
+{
+	uint16_t regval;
+
+	/* The charge current servo level: (icharge_dac + 1) × 1mV/RSNSB */
+	regval = LTC4162L_MULT_FRAC(val, dev->rsnsb, 100000u);
+	regval /= 10000u;
+
+	if (regval)
+		--regval;
+
+	if (regval > 31)
+		return -EINVAL;
+
+	return ltc4162l_reg_write(dev, LTC4162L_CHARGE_CURRENT_SETTING, regval);
+}
+
+/**
+ * @brief Get icharge value
+ * @param dev - The LTC4162L device structure.
+ * @param val - The icharge value.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_get_icharge(struct ltc4162l_desc *dev, uint16_t *val)
+{
+	uint16_t regval;
+	int ret;
+
+	ret = ltc4162l_reg_read(dev, LTC4162L_ICHARGE_DAC, &regval);
+	if (ret)
+		return ret;
+
+	regval = no_os_field_get(LTC4162L_ICHARGE_DAC_MASK, regval);
+
+	/* The charge current servo level: (icharge_dac + 1) × 1mV/RSNSB */
+	++regval;
+
+	*val = 10000u * LTC4162L_MULT_FRAC(regval, 100000u, dev->rsnsb);
+
+	return 0;
+}
+
+/**
+ * @brief Get the vcharge setting
+ * @param dev - The LTC4162L device structure.
+ * @param val - The vcharge setting.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ltc4162l_get_vcharge_setting(struct ltc4162l_desc *dev, uint16_t *val)
+{
+	uint16_t regval;
+	int ret;
+	uint32_t voltage;
+	uint16_t cell_count;
+	uint16_t chem_type;
+
+	ret = ltc4162l_reg_read(dev, LTC4162L_VCHARGE_SETTING, &regval);
+
+	if (ret)
+		return ret;
+
+	regval = no_os_field_get(LTC4162L_VCHARGE_SETTING_MASK, regval);
+
+	ret = ltc4162l_get_cell_count(dev, &cell_count);
+	if (ret)
+		return ret;
+
+	ret = ltc4162l_get_chem_type(dev, &chem_type);
+	if (ret)
+		return ret;
+
+	switch (dev->type) {
+	case ID_LTC4162L:
+		voltage = 3812500 + (regval * 12500);
+		voltage *= cell_count;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4162F:
+		voltage = 3412500 + (regval * 12500);
+		voltage *= cell_count;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4162S:
+		voltage = 6000000 + (regval * 28571);
+		voltage *= cell_count/2;
+		*val = voltage;
+
+		return 0;
+	case ID_LTC4015:
+		if (chem_type <= 3)
+			voltage = 3812500 + (regval * 12500);
+		else if (chem_type > 3 || chem_type < 7)
+			voltage = 3412500 + (regval * 12500);
+		else
+			voltage = 2000000 + (regval * 9523);
```

Comment:
ditto

Reply:
moved to chip_info function and created new computation function